File: torture-chamber.cpp
#include <bits/stdc++.h>
#include <unordered_set>
#include <stdio.h>

using namespace std;

#define f first
#define s second
#define ll long long
#define ull unsigned long long
#define PI 3.141592653589793
#define vi vector<int>
#define vll vector<long long>
#define p pair<int,int>
const int MOD = 1000000007;
inline double degcos(double ang) {return cos(ang*PI/180);}
inline double degsin(double ang) {return sin(ang*PI/180);}
#define rep(i,j,k,in) for (int i = j; i < k; i += in)


/*---END OF TEMPLATE---*/


vector<bool> segmentedSieve(ll L, ll R) {
    ll limit = sqrt(R);
    vector<bool> mark(limit + 1, false);
    vector<ll> primes;
    for (ll i = 2; i <= limit; ++i) {
        if (!mark[i]) {
            primes.push_back(i);
            for (ll j = i * i; j <= limit; j += i) //all multiples before it have been accounted for
                mark[j] = true;
        }
    }

    vector<bool> isPrime(R - L + 1, true);
    for (ll i : primes)
        for (ll j = max(i * i, (L + i - 1) / i * i); j <= R; j += i) isPrime[j - L] = 0;
    if (L == 1 && isPrime[0]) isPrime[0] = 0;
    return isPrime;
}

int main() {
    ll a,b;
    cin >> a >> b;
    ll tot = 0;
    for (bool b : segmentedSieve(a,b-1)) {
        if (b) ++tot;
    }
    cout << tot << endl;

}

File: torture-chamber.py
from math import sqrt
def segmentedSieve(L,R):
    limit = int(sqrt(R))
    mark = [False for i in range(limit+1)]
    primes = []
    for i in range(2,limit+1):
        if not mark[i]:
            primes.append(i)
            for j in range(i*i,limit+1,i):
                mark[j] = True
    isPrime = [True for i in range(R-L+1)]
    for i in primes:
        for j in range(max(i * i, (L + i - 1) // i * i),R+1,i):
            isPrime[j-L] = 0
    if L == 1 and isPrime[0]:
        isPrime[0] = False
    return isPrime

a = int(input())
b = int(input())
tot = 0
for b in segmentedSieve(a,b-1):
    if b:
        tot += 1
print(tot)
